Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Views/ViewBase.cs

```diff
@@ -0,0 +1,291 @@
+﻿using System;
+using System.Collections.Generic;
+
+namespace UnityEngine.InputSystem.Editor
+{
+	internal interface IViewStateSelector<out TViewState>
+	{
+		bool HasStateChanged(InputActionsEditorState state);
+		TViewState GetViewState(InputActionsEditorState state);
+	}
+
+	internal interface IView
+	{
+		void UpdateView(InputActionsEditorState state);
+		void DestroyView();
+	}
+
+	internal abstract class ViewBase<TViewState> : IView
+	{
+		public event Action<ViewBase<TViewState>> OnClosing;
+
+		protected ViewBase(StateContainer stateContainer)
+		{
+			this.stateContainer = stateContainer;
+			m_ChildViews = new List<IView>();
+		}
+
+		protected void OnStateChanged(InputActionsEditorState state)
+		{
+			UpdateView(state);
+		}
+
+		public void UpdateView(InputActionsEditorState state)
+		{
+			if (m_ViewStateSelector == null)
+			{
+				Debug.LogWarning($"View '{GetType().Name}' has no selector and will not render. Create a selector for the " +
+				                 $"view using the CreateSelector method.");
+				return;
+			}
+
+			if (m_ViewStateSelector.HasStateChanged(state) || m_IsFirstUpdate)
+				RedrawUI(m_ViewStateSelector.GetViewState(state));
+
+			m_IsFirstUpdate = false;
+			foreach (var view in m_ChildViews)
+			{
+				view.UpdateView(state);
+			}
+		}
+
+		public TView CreateChildView<TView>(TView view) where TView:IView
+		{
+			m_ChildViews.Add(view);
+			return view;
+		}
+
+		public void Close()
+		{
+			OnClosing?.Invoke(this);
+		}
+
+		public void DestroyView<TView>(TView view) where TView : IView
+		{
+			if (view == null)
+				return;
+
+			m_ChildViews.Remove(view);
+			view.DestroyView();
+		}
+
+		public void Dispatch(Command command)
+		{
+			stateContainer.Dispatch(command);
+		}
+
+		public virtual void RedrawUI(TViewState viewState)
+		{
+			
```

Comment:
Intentionally empty?

Reply:
I didn't want every view to have to implement RedrawUI, but it seems they all do now anyway, so made this one abstract.